metrics/metrics.go

```diff
@@ -73,25 +80,47 @@ type CSIMetricsManager interface {
 	StartMetricsEndpoint(metricsAddress, metricsPath string)
 }
 
+// MetricsManagerOption is used to pass optional configuration to a
+// new metrics manager.
+type MetricsManagerOption func(*csiMetricsManager)
+
+// WithSubsystem overrides the default subsystem name.
+func WithSubsystem(subsystem string) MetricsManagerOption {
```

Comment:
Do we want the subsystem to be configurable or should we just have 2 constructor APIs, one for sidecars and one for plugins?

What happens if someone decides to input their own subsystem name?

Reply:
> Do we want the subsystem to be configurable or should we just have 2 constructor APIs, one for sidecars and one for plugins?

Configurable, please. In PMEM-CSI, CreateVolume is passed from sidecar to controller to a driver on a node. All three have metrics gathering enabled. The same argument about counting the same operation multiple times also applies there, i.e. all three must use different subsystem. I ended up using "csi_sidecar", "csi_controller", "csi_node". 

> What happens if someone decides to input their own subsystem name?

My assumption was that a CSI driver is free to choose the names for itself. But I am not sure anymore, because the names are the same regardless of their origin. But this is general issue with Prometheus, right? One component can provide a metric "foobar" with a certain semantic and another component some other metric "foobar" with a completely different semantic.

Reply:
Can we have both? Standard constructors for the 2 most common use cases: a sidecar or a plugin.  And a 3rd customizable option if the driver is doing some advanced stuff.

Standardizing on a component name for plugins would be really nice so we can write generic collectors that work for all csi drivers.

Reply:
> Can we have both? Standard constructors for the 2 most common use cases: a sidecar or a plugin. And a 3rd customizable option if the driver is doing some advanced stuff.

Okay.

> Standardizing on a component name for plugins would be really nice so we can write generic collectors that work for all csi drivers.

With collectors, do you mean Prometheus jobs for scraping? Those don't depend on the metrics names. More problematic for scraping is how pods are labeled: there's no established convention for scraping multiple containers in a pod. I made a proposal in https://github.com/helm/charts/issues/22899 but got no response.

Standardizing the names becomes more interesting when running queries or creating dashboards.

Reply:
Yes sorry I meant dashboards.